### ISSUE TITLE
[Feature Request]:  Please support Wayland by default in Electron on Linux

### ISSUE THREAD

**Issue by GerryT11** [electron/electron#41551]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

26.2.4

### What operating system are you using?

Ubuntu

### Operating System Version

6.2.0-39-generic #40-Ubuntu SMP PREEMPT_DYNAMIC Tue Nov 14 14:18:00 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### What arch are you using?

x64

### Expected Behavior

As of today, Electron applications do not support/enable Wayland out of the box, but rather run under XWayland. Please support Wayland by default in Electron on Linux. Are there any blocker issues to enable Wayland by default?

For example, I had a discussion with RStudio developers who decided to wait enabling Wayland by default until Wayland is supported by default out-of-the-box in Electron. https://github.com/rstudio/rstudio/issues/11962#issuecomment-1984134246

### Actual Behavior

.

**Comment by ryanabx** [comment]
I agree -- I feel like it's well past time for ozone-platform to be auto on electron. I don't know necessarily what the hold up is upstream (chromium), though if someone knows please let me know, I've been trying to figure it out for a while now.

**Comment by vially** [comment]
#35630 would address this, [but it was argued](https://github.com/electron/electron/pull/35630#issuecomment-1254037947) that Electron shouldn't deviate from Chromium on this.

**Comment by ryanabx** [comment]
Yeah, it'd of course be nice to follow upstream on this, but I still am left wondering when chromium will support Wayland/auto by default?

Chromium ought to have a roadmap or an issue tracker for Wayland enablement.

**Comment by veodko** [comment]
What was a problem on my side anyway was the fact that electron prebuilt binary is linked to a ton of X11 libraries that are not available when operating on a wayland only distribution obviously (custom yocto/buildroot builds). No matter what I set in `ELECTRON_OZONE_PLATFORM_HINT`, I couldn't run my electron app on wayland because of missing libraries like `libXrandr.so` or `libX11.so`. Even though they are probably not used on wayland, they are still linked to the binary and required.

How do you go around this? I just copied a bunch of libs and passed them using `LD_LIBRARY_PATH` which finally made it work, but it's a cumbersome solution. X11 should die a long time ago.

**Comment by banekondic1996** [comment]
It's quite easy. Electron is supported by globalists, used by Microsoft, and Microsoft doesn't want much of a Linux support. I didn't have much luck with Electron, so i use NW.JS which does support wayland, but Electron has more features and is better supported in some ways

**Comment by v3ss0n** [comment]
> What was a problem on my side anyway was the fact that electron prebuilt binary is linked to a ton of X11 libraries that are not available when operating on a wayland only distribution obviously (custom yocto/buildroot builds). No matter what I set in `ELECTRON_OZONE_PLATFORM_HINT`, I couldn't run my electron app on wayland because of missing libraries like `libXrandr.so` or `libX11.so`. Even though they are probably not used on wayland, they are still linked to the binary and required.
> 
> How do you go around this? I just copied a bunch of libs and passed them using `LD_LIBRARY_PATH` which finally made it work, but it's a cumbersome solution. X11 should die a long time ago.

X11 won't die  - Wayland would die first and replace by something else . Wayland DEVs's made so many decisions that screws wayland. 

Here are problems of wayland and many of them wont be fixed , because it against the wayland core develoeprs's belives and ideologies just like religious fanatics. 

https://gist.github.com/probonopd/9feb7c20257af5dd915e3a9f2d1f2277

**Comment by ryanabx** [comment]
We're getting pretty far off topic with that previous comment. The reality is that linux is moving in the direction of wayland, because it's what most developers are supporting nowadays. Make something worthwhile, and linux will move that way instead. There's no conspiracy, nothing is being forced onto people, just open source being open source

**Comment by fxzxmicah** [comment]
> > What was a problem on my side anyway was the fact that electron prebuilt binary is linked to a ton of X11 libraries that are not available when operating on a wayland only distribution obviously (custom yocto/buildroot builds). No matter what I set in `ELECTRON_OZONE_PLATFORM_HINT`, I couldn't run my electron app on wayland because of missing libraries like `libXrandr.so` or `libX11.so`. Even though they are probably not used on wayland, they are still linked to the binary and required.
> > How do you go around this? I just copied a bunch of libs and passed them using `LD_LIBRARY_PATH` which finally made it work, but it's a cumbersome solution. X11 should die a long time ago.
> 
> X11 won't die - Wayland would die first and replace by something else . Wayland DEVs's made so many decisions that screws wayland.
> 
> Here are problems of wayland and many of them wont be fixed , because it against the wayland core develoeprs's belief and ideologies just like religious fanatics.
> 
> https://gist.github.com/probonopd/9feb7c20257af5dd915e3a9f2d1f2277
> 
> Don't get me wrong. There are many problems at X11 , which could be solved if there is a new X11 that might deviate from X11 protocol but backward compatible - instead of a rewrite filled with bad decision that breaks so many things - by design.

But your fantasy didn't come true. In addition, the default support for wayland does not mean that x11 is removed.

**Comment by fxzxmicah** [comment]
`use_ozone=true ozone_auto_platforms=false ozone_platform="wayland" ozone_platform_wayland=true`
Build with these parameters to get a wayland-based chromium.

**Comment by fxzxmicah** [comment]
> [#35630](https://github.com/electron/electron/pull/35630) would address this, [but it was argued](https://github.com/electron/electron/pull/35630#issuecomment-1254037947) that Electron shouldn't deviate from Chromium on this.

In theory, we should follow upstream, but in practice, chromium is just one software, and its unsupported state can be changed with a single command line argument. Electron represents a myriad of software, which requires the user to set myriad command-line arguments. In this case, the benefits of following upstream are limited.

**Comment by willmtemple** [comment]
I just wanted to add some evidence to this issue of _why_ this would be useful. It's not just an implementation detail, it negatively impacts how Electron apps render on some Linux desktop environments. In particular, on GNOME when using UI scaling, especially fractional scaling, it causes text to look blurry. Here's a side by side image of Discord loaded in Thorium using `--ozone-platform=wayland` (left) and Discord packaged in an Electron app (right):

![Image](https://github.com/user-attachments/assets/51f209ef-6080-4706-9bf8-ed30fb7eadee)

You can see that the text on the left (Thorium) is crisp, but on the right (Electron) the whole application is blurry and this really stands out on text. This affects every electron app on my system and it's been that way for _years_.

Running Discord via `flatpak run --socket=wayland com.discordapp.Discord --enable-features=UseOzonePlatform --ozone-platform=wayland` fixes the text issue. However, it introduces other issues of its own: no tray icon and some weird rendering artifacts in the splash screens.

**Comment by dev-bre** [comment]
is there any intention on adding native support for wayland? I malso noticed that window resize and zoom factor setting do not work either when wayland is enabled.

**Comment by ryanabx** [comment]
> is there any intention on adding native support for wayland? I malso noticed that window resize and zoom factor setting do not work either when wayland is enabled.

That's likely a chromium issue, but not sure

**Comment by eszlari** [comment]
https://chromium-review.googlesource.com/c/chromium/src/+/6771245

```
Enable OverrideDefaultOzonePlatformHintToAuto feature

This is required for cl/784605417 which fully launches the experiment.

Change-Id: [If722ab1b2a8f6b1004b4baf22834f8f026e8eb5c](https://chromium-review.googlesource.com/q/If722ab1b2a8f6b1004b4baf22834f8f026e8eb5c)
Bug: [40083534](https://issues.chromium.org/issues/40083534)
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6771245
Commit-Queue: Thomas Anderson <thomasanderson@chromium.org>
Reviewed-by: Kramer Ge <fangzhoug@chromium.org>
Reviewed-by: Nico Weber <thakis@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1489138}
```

**Comment by Thatoo** [comment]
@eszlari could you explain what does that mean? Does that mean this option has been merged into chromium now and next release of electron based on this new version of chromium will solve our issues?

**Comment by eszlari** [comment]
Upstream Chromium wants to enable Wayland by default in the development builds as an experiment, but it will take some time until this reaches the stable channel, see: https://chromiumdash.appspot.com/schedule

**Comment by VictorVow** [comment]
[Chromium 141 and up will use Wayland by default](https://issues.chromium.org/issues/40083534#comment593) it looks like.

**Comment by ryanabx** [comment]
We are using Chromium 141 in electron main branch now. Anybody who uses electron -- can we test to see if electron apps using chromium 141 have wayland on by default?

Edit: I ran 39.0.0-alpha.1 which should have chromium 141, and it's still defaulting to X11:

<img width="708" height="444" alt="Image" src="https://github.com/user-attachments/assets/e6653a15-0103-4299-9670-086e914c1cdf" />

I think we're going to have to change something in electron to pick up this behavior so that we don't have to use `--ozone-platform=wayland`

**Comment by ryanabx** [comment]
> [Chromium 141 and up will use Wayland by default](https://issues.chromium.org/issues/40083534#comment593) it looks like.

Actually, it does look like Chromium 140 is wayland by default too!

**Comment by diegorondini** [comment]
> https://chromium-review.googlesource.com/c/chromium/src/+/6771245
> 
> ```
> Enable OverrideDefaultOzonePlatformHintToAuto feature
> 
> This is required for cl/784605417 which fully launches the experiment.
> 
> Change-Id: [If722ab1b2a8f6b1004b4baf22834f8f026e8eb5c](https://chromium-review.googlesource.com/q/If722ab1b2a8f6b1004b4baf22834f8f026e8eb5c)
> Bug: [40083534](https://issues.chromium.org/issues/40083534)
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6771245
> Commit-Queue: Thomas Anderson <thomasanderson@chromium.org>
> Reviewed-by: Kramer Ge <fangzhoug@chromium.org>
> Reviewed-by: Nico Weber <thakis@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1489138}
> ```

This was actually immediately reverted (as can be seen in the last gerrit change log):
https://chromium-review.googlesource.com/c/chromium/src/+/6772226

The change actually went in August 4, 2025, which is included in Chromium 141.0.7340.0+:
https://github.com/chromium/chromium/commit/8f2c7d1c8585203b30d8f8c893f24a269a3f5417

**Comment by pounard** [comment]
Since electron 38 upgrade (element-desktop, signal) it doesn't honnor system-wide configuration using environment variables anymore, and doesn't seem to accept any command line option anymore either. Fractional scaling is broken, is there anything I missed?

**Comment by mitchchn** [comment]
Electron runs in Wayland by default as of Electron 38.2 and newer. You can opt out with a command line flag but there is no longer an environment variable:
https://www.electronjs.org/blog/electron-38-0#removed-electron_ozone_platform_hint-environment-variable

**Comment by pounard** [comment]
Sad news, the environment variable was a convenient way to set the setting system-wide without having to edit or create new .desktop file for each app using it.

**Comment by nikwen** [comment]
Closing this as per @mitchchn's comment.

**Comment by pounard** [comment]
I don't want to be rude, at anytime, because Electron is quite a complex and good product, but it seems that this time, you were maybe too eager to get rid of deprecated behaviors such as environment variables configuration, and yet the feature end up really working only 2 point releases after that, in the meantime, a lot of products using yours were broken. Some will not upgrade before again yet a few month. You moved too quickly, and it broke a lot of apps. Please, be more careful with users the next time.

Edit: don't take it wrong, I do really appreciate electron when it brings new features such as wayland support per defaut, it's really a good thing for all hidpi screens owner, thanks.

**Comment by debuxxed** [comment]
doesn't work for me, as an app (Vesktop) that uses Electron 39, still runs under xwayland (Niri compositor)

**Comment by nikwen** [comment]
>  You moved too quickly, and it broke a lot of apps.

Sorry for that! We try to be careful and never break behavior on purpose.

> an app (Vesktop) that uses Electron 39, still runs under xwayland

The app itself can choose to run under XWayland if it wants to. There are some features that are not supported by Wayland that are supported by X11, so some apps might make that choice.
